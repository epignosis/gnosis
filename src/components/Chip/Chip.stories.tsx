import React from "react";
import { Story } from "@storybook/react";
import Chip, { ChipProps } from "./Chip";

export default {
  component: Chip,
  title: "Components/Chip",
  argTypes: {
    onClose: { action: "clicked" },
    size: {
      control: {
        type: "select",
        options: ["md", "lg"],
      },
    },
  },
};

export const Default: Story<ChipProps> = (args) => <Chip {...args} />;

Default.args = {
  size: "md",
  style: { backgroundColor: "#5c5c5c", color: "#fff" },
  children: "This is a chip",
};
