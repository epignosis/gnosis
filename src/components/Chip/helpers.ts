import { MutableRefObject, useLayoutEffect, useState } from "react";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const useIsOverflowX = (overflowRef: MutableRefObject<any>): boolean => {
  const [isOverflow, setIsOverflow] = useState<boolean>(false);

  useLayoutEffect(() => {
    const el = overflowRef.current;

    const trigger = (): void => {
      setIsOverflow(el.offsetWidth < el.scrollWidth);
    };

    if (el) {
      if ("ResizeObserver" in window) {
        new ResizeObserver(trigger).observe(el);
      }

      trigger();
    }
  }, [overflowRef]);

  return isOverflow;
};
