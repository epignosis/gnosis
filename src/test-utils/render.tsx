import React, { FC, ReactElement, ComponentType } from "react";
import { render, RenderResult, RenderOptions } from "@testing-library/react";
import ThemeProvider from "../components/ThemeProvider/ThemeProvider";

const AllTheProviders: FC = ({ children }) => <ThemeProvider>{children}</ThemeProvider>;

const customRender = (ui: ReactElement, options: RenderOptions): RenderResult =>
  render(ui, { wrapper: AllTheProviders as ComponentType, ...options });

export * from "@testing-library/react";

export { customRender as render };
