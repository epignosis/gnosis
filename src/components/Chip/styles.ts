import { css, Theme, SerializedStyles } from "@emotion/react";
import { Size } from "./Chip";

const chipHeight = {
  md: "2.5rem",
  lg: "3rem",
};

export const chip = ({ typeScaleSizes }: Theme, { size }: { size: Size }): SerializedStyles => {
  const fontSizes = {
    md: typeScaleSizes.sm,
    lg: typeScaleSizes.lg,
  };

  return css`
    display: inline-flex;
    align-items: center;
    font-size: ${fontSizes[size]};
    ${size && `height: ${chipHeight[size]}`};
    padding: 0 0.75rem;
    border-radius: 5px;

    button {
      margin-inline-end: 0.5rem;
      display: inline-flex;
      color: white;
      background: transparent;
      border: none;
      padding: 0;
    }
  `;
};
