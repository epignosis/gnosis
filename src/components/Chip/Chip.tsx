import React, { FC, MouseEvent } from "react";
import { SerializedStyles } from "@emotion/react";
import { CloseSVG } from "../../icons/";
import { chip } from "./styles";

export type Size = "md" | "lg";

export type ChipProps = React.HTMLAttributes<HTMLDivElement> & {
  onClose?: (e: MouseEvent) => void;
  size?: Size;
  children: string;
};

const Chip: FC<ChipProps> = ({
  size = "md",
  onClose,
  children,
  style = { backgroundColor: "#5c5c5c", color: "#fff" },
  ...rest
}) => (
  <div css={(theme): SerializedStyles => chip(theme, { size })} style={style} {...rest}>
    {onClose && (
      <button onClick={onClose} aria-label={`Remove ${children}`} style={{ color: style.color }}>
        <CloseSVG height={16} />
      </button>
    )}

    {children}
  </div>
);

export default Chip;
