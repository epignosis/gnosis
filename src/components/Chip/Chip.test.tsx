import React from "react";
import userEvent from "@testing-library/user-event";
import faker from "faker";
import Chip from "./Chip";
import { render, screen } from "@test-utils/render";

describe("<Chip />", () => {
  it("renders without close button", () => {
    const chipTxt = faker.lorem.word();

    render(<Chip>{chipTxt}</Chip>);
    const tag = screen.getByText(chipTxt);

    expect(tag).toHaveTextContent(chipTxt);
  });

  it("renders with close button", () => {
    const chipTxt = faker.lorem.word();
    const mockFn = jest.fn();

    render(<Chip onClose={mockFn}>{chipTxt}</Chip>);

    const tag = screen.getByText(chipTxt);
    const closeBtn = screen.getByLabelText(`Remove ${chipTxt}`);

    userEvent.click(closeBtn);

    expect(tag).toHaveTextContent(chipTxt);
    expect(mockFn).toHaveBeenCalledTimes(1);
  });

  it("matches snapshot", () => {
    const { container } = render(
      <Chip
        size="md"
        className="my-class"
        style={{ backgroundColor: "purple", color: "white" }}
        onClose={jest.fn()}
      >
        This is a chip
      </Chip>,
    );

    expect(container).toMatchSnapshot();
  });
});
