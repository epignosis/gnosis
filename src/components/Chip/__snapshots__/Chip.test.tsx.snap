// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Chip /> matches snapshot 1`] = `
<div>
  <div
    class="my-class css-lo603d-chip-chip"
  >
    This is a chip
  </div>
</div>
`;
